src/nccl_ofi_net.c

```diff
@@ -1050,49 +1064,130 @@ static inline ncclResult_t nccl_ofi_progress(nccl_ofi_t *nccl_ofi_comp)
 }
 
 /*
- * @brief	Checks if the underlying hardware type matches the input platform type.
+ * @brief	Provides EC2 platform type as reported by the
+ * 		first line of
+ *		/sys/devices/virtual/dmi/id/product_name.
+ *		Users of this API *should* free the buffer when a
+ *		Non-NULL string is returned.
  *
- * @return	-1, on error
- *		1, if it matches
- *		0, if it does not match
+ * @return	NULL, on allocation and file system error
+ * 		EC2 platform type, on success
  */
-static int platform_type_matches(char *platform_type)
+static char* get_platform_type()
 {
 	char file[] = "/sys/devices/virtual/dmi/id/product_name";
 	FILE *fd = NULL;
-	int len = strlen(platform_type);
-	int idx = 0, ret = -1;
 	char ch;
+	size_t len = 0;
+	size_t platform_type_len = 64;
+	char *platform_type = NULL;
 
 	fd = fopen(file, "r");
 	if (fd == NULL) {
 		NCCL_OFI_WARN("Error opening file: %s", file);
-		return -1;
+		goto error;
 	}
 
-	/* Read first line of the file */
+	platform_type = (char *)malloc(sizeof(char)*platform_type_len);
+	if (platform_type == NULL) {
+		NCCL_OFI_WARN("Unable to allocate platform type");
+		goto error;
+	}
+
+	/* Read first line of the file, reallocing the buffer as necessary */
 	while ((feof(fd) == 0) && (ferror(fd) == 0) && ((ch = fgetc(fd)) != '\n')) {
+		platform_type[len++] = ch;
+		if (len > platform_type_len) {
+			platform_type = realloc(platform_type, len + platform_type_len);
+		}
+	}
+	platform_type[len] = '\0';
+
+	NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET, "Read %d bytes. EC2 platform type is %s", len, platform_type);
+
+	fclose(fd);
+	return platform_type;
+
+error:
+	if (platform_type)
+		free(platform_type);
+	if (fd)
+		fclose(fd);
+	return platform_type;
+}
 
-		/* Report mismatch if characters read are more than string length */
-		if (idx == len) {
-			ret = 0;
+/*
+ * @brief	Returns static topology filename for given platform type, if found
+ *
+ * @input	Platform type
+ *
+ * @return	NULL, if no topology found
+ * 		Topology filename, if match found
+ */
+static const char* get_static_topology_file(char *platform_type)
```

Comment:
You know, we should just convert this file to C++ so we never have to write this kind of thing again :)

(obviously out of scope for this review)